[benchmark] Groundwork for Robust Measurements

palimondo opened the pull request:
These improvements to benchmarking infrastructure include bug fixes, various simplifications, some clean up refactoring and preliminary support for Python 3. It is pawing way for more robust measurements and has been split-off from #26462.

Please review by [individual commits](https://github.com/apple/swift/pull/30158/commits/ac294f39867559ea8c1373d5451e01595d907a83), which are self-contained.

palimondo committed:
[benchmark] Fix parsing delta zeroed metadata

palimondo committed:
[benchmark] BenchmarkDriver min-samples & metadata

Support for invoking benchmark drivers with min-samples and gathering environmental metadata.

palimondo committed:
[benchmark] PerformanceTestResults merge samples

Adjusted how merged PerformanceTestResults track the number of underlying samples when using quantile subsampling.

palimondo committed:
[benchmark] Don’t justify last column in reports

In text reports, don’t justify the last columns with unnecessary spaces.

palimondo committed:
[benchmark] Retire old num-iters in verbose format

After commit 331c0bf772a4626aa1f8e1f8aae7629e10d350db from a year ago, all samples from the same run have the same num-iters.

palimondo committed:
[benchmark] [Gardening] Fix assertEqual naming

palimondo committed:
[benchmark] Add tests for all_samples

palimondo committed:
[benchmark] Refactor: store num_iters on PTS

Store the number of iterations averaged in each sample on the PerformanceTestSamples.

palimondo committed:
[benchmark] Refactor: remove class Sample

Removed Sample class, that was previously holding num_iters and the ordinal number of the sample.

palimondo committed:
[benchmark] Refactor: simpler exclude_outliers

palimondo committed:
[benchmark] More immutable PerformanceTestSamples

Removed the ability to add individual samples to PTS.
PerformanceTestSamples are technically not really immutable, because of exclude_outliers.

palimondo committed:
[benchmark] Fix crash in oversampled results

Gracefully handle the parsing of oversampled values in critical configuration, when the sampling error causes an ommision of certain quantiles from the report.

palimondo committed:
[benchmark] Report ventiles for dubious results

For dubious result comparisons, print out empirical sample distribution (ventiles) to enable humans to reach informed decisions about these performance changes.

palimondo committed:
[benchmark] Refactor: remove redundant _runtimes

palimondo committed:
[benchmark] Collate metadata when merging PTRs

Handle optional `--meta` data in `merge` of `PerformanceTestResults`. Pick minimum from memory pages, sum the number of involuntrary contex switches and yield counts.

palimondo committed:
[benchmark] Keep merged independent run samples

When merging `PerformanceTestResults`s keep the original `PerformanceTestSample`s from all independent runs. These will be used to choose the most stable (least variable) location estimate for the `ResultComparison` down the road.

palimondo committed:
[benchmark] Override all_samples on merged results

To save on memory used by merged `PerformanceTestResult`s, the rarely used `PerformanceTestSample.all_samples` can gather the samples on demand from the result’s `independent_runs` instead of keeping another copy.

palimondo committed:
[benchmark] Python 3 Support

palimondo committed:
Revert "[NFC][Python: black] Reformatted the benchmark Python sources using utils/python_format.py."

palimondo committed:
[benchmark] Dubious indicator for changes only

Display the `(?)` indicator for dubious result only for changes, never for the unchanged results.
Refactored `ResultComparison.init` with simplyfied range check.

palimondo commented:
@broadwaylamb I’ve tried to incorporate your work from #30085 here, but I chose to refactor few parts for more idiomatic list comprehensions where possible. I didn’t touch the additional .py files that aren’t part of main benchmarking infrastructure that I’m familiar with. Could you please have a look if I missed something? See commit ec64535e00b1e16a122d186b2c2182b7cbe3adad

palimondo commented:
@eeckstein please review

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test os x

broadwaylamb commented on benchmark/scripts/Benchmark_Driver:
```diff
@@ -436,122 +373,95 @@ class BenchmarkDoctor(object):
+            [(result.samples.min - correction) for correction, i_series in
+             [(correction, BenchmarkDoctor._select(measurements, num_iters=i))
+              for i, correction in [(i, setup // i) for i in [1, 2]]]
+             for result in i_series])
```
I personally find deeply nested list comprehensions significantly harder to read compared to for loops. Especially here, where you have to propagate the `i` and `correction` variables, which would not be necessary if we were using the for loop:

```python
runtimes = []
for i in range(1, 3):
    correction = setup / i
    i_series = BenchmarkDoctor._select(measurements, num_iters=i)
    for result in i_series:
        runtimes.append(result.samples.min - correction)
runtime = min(runtimes)
```

palimondo replied:
This really comes down to personal taste. I have trouble reading procedural code with side effects. I’ve [answered below](https://github.com/apple/swift/pull/30158#commitcomment-37585480) with link to direct comparison of the Python 3 changes.

palimondo replied:
@eeckstein Do you have a preference here?

eeckstein replied:
I find @broadwaylamb's version _much_ easier to understand. That's probably because I'm not a python expert (which is probably the case for many our contributors).

In general, making the code easy to understand should be a central goal.

palimondo replied:
Refactored in bf06df6.

palimondo commented:
@swift-ci clean smoke test os x

broadwaylamb commented on benchmark/scripts/test_Benchmark_Driver.py:
```diff
@@ -279,246 +262,219 @@ def setUp(self):
-            with open(log_file, "rU") as f:
+            self.assertEqual(out.getvalue(),
+                             'Logging results to: ' + log_file + '\n')
+            with open(log_file, 'r') as f:
```
The `'rU'` mode's behavior is the default in Python 3, and explicitly setting this mode is deprecated (a warning is printed to stderr).

However, I'm not sure this change won't affect Python 2 compatibility. In #30085 I've defensively added a version check:

```python
if sys.version_info < (3, 0):
    openmode = "rU"
else:
    openmode = "r"
with open(log_file, openmode) as f:
    # ...
```

Are you sure this is okay to just use `'r'` here?

palimondo replied:
All test passed with python3 in shebang, so I think the answer is yes.

broadwaylamb replied:
Have you tested it on Windows, where the newlines are not represented with `'\n'`?

palimondo replied:
I did not. @eeckstein @compnerd Do we support swift benchmarking on Windows already?

palimondo replied:
@compnerd I tried triggering “@swift-ci please test Windows platform” in hope that it runs also the python unit tests for benchmarks. Is that a correct assumption?

palimondo replied:
OK, I think it does not at the moment, because of https://github.com/apple/swift/blob/master/test/benchmark/benchmark-scripts.test-sh#L2

palimondo commented:
@shahmishal Smoke test on OS X keeps crashing with 'clang-7: error: unable to execute command: Killed: 9’, previously it also said the git repo was corrupted. Do you know what is wrong?

shahmishal commented:
Can you try again? I took this node offline.

palimondo commented:
@swift-ci please smoke test os x

palimondo commented:
@swift-ci please smoke test os x

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please python lint

palimondo commented:
@swift-ci please python lint

palimondo commented:
@swift-ci please smoke test os x

palimondo commented:
@swift-ci please smoke test os x

palimondo commented:
@shahmishal The Linux smoke test failed with strange error. Definitely unrelated to my changes. Could you please have a look? Do I need to run these with some *clean* modifier for @swift-ci?

shahmishal commented:
@swift-ci please clean smoke test Linux

eeckstein commented:
This is a huge PR. It seems to me that there are completely different things mixed up here (refactoring, functional changes). Though it consists of smaller commits, I'd prefer to have smaller PRs which can land over time. This makes it easier to test and revert if something breaks.

In general, I'm kind of missing the motivation for many of those changes, e.g. which problem do the changes solve? It can be as simple as "this refactoring makes the code easier to read" up to "this fixes bug X or problem Y".

It feels like putting a lot of effort to go from A to B. But why is B better than A?

palimondo commented:
@eeckstein This PR boils down to this LOC stat: +1,939 **−2,452**.
It does more with less code. As usual, all is covered by unit tests and explained in commit comments.

Since this PR is spun-off from #26462, except few final commits that are “flipping the switch” on new measurement methodology (as @gottesmm put it).

The objective of [WIP] #26462 is to find more statistically rigorous method of detecting performance changes in our benchmarks, while minimizing type [I and type II errors](https://en.wikipedia.org/wiki/Type_I_and_type_II_errors). Our current measurement process still suffers from occasional false positives (type I errors). 

The means for improving that situation is to make collection of many samples from multiple independent runs and use more sophisticated statistical analysis. The long term opportunity, if the new method from _Robust Measurements_ pans out, is to speed up the whole process by running benchmarks in parallel (this would be too risky with status quo, because we work with aggregates of averages and do not monitor the environment properly).

If it eases things for you, I could split this PR into several more, but it would IMHO only increase your review load for no reason.

palimondo committed:
[Gardening] Refactored runtimes correction for SO

Replaced list comprehension that computes the minimum from runtimes corrected for setup overhead with a procedural style that is easier to understand.

palimondo commented:
@swift-ci python lint

palimondo commented:
@swift-ci python lint

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@eeckstein One more thing! Maybe it wasn’t clear, but this builds on open PR #30149 which reverts the automated formatting from #29719. That is the only reason why it now looks like it is touching 17 files in benchmarks. That is all in the first commit. If we merge #30149, as I think we should, it would disappear from here and it’ll all look much simpler!

palimondo commented:
@shahmishal I don’t get this… OS X smoke test reports:
````
20:02:57 Failing Tests (1):
20:02:57     Swift(macosx-x86_64) :: Python/python_lint.swift
````
But I explicitly invoked the separate Python lint check from @swift-ci which passed. I see no issues running python_lint.py locally either.

palimondo commented:
@swift-ci please clean smoke test OS X

palimondo commented:
@swift-ci please clean smoke test OS X

palimondo commented:
@swift-ci please test Windows platform

palimondo commented:
@swift-ci please test Windows platform

palimondo commented:
@eeckstein I have [replied above](https://github.com/apple/swift/pull/30158#issuecomment-594076344) regarding the motivation for this PR. Did that answer your questions? Do you have objections I should address to move this forward?

palimondo commented:
@swift-ci please test Windows platform

palimondo commented:
@swift-ci please test Windows platform

palimondo commented:
@swift-ci please clean smoke test OS X

palimondo commented:
@shahmishal Could you please have a look at the Windows machine and also on the python lint issue above?

shahmishal commented:
@compnerd windows nodes are having issues, can you look at them?

shahmishal commented:
@palimondo do you have the full out from the test failure? 

```
Swift(macosx-x86_64) :: Python/python_lint.swift
```

palimondo commented:
@shahmishal Not anymore. Sorry I got inpatient and triggered a re-test a moment ago. Either it’ll pass or there’ll be a fresh log ready in about an hour? My apologies!

palimondo commented:
@swift-ci please test Windows platform

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
